nspawn: Do not `setsid` or surrender controlling terminal with `--console=pipe --as-pid2`

snarkmaster opened the pull request:
As far as I understand, both `setsid` calls, and the `TIOCNOTTY` ioctl were introduced in support of the "nspawn allocates a PTY" use-case.

- https://github.com/systemd/systemd/commit/a258bf2648fa5e3005b96203c60b9751807bacf0
- https://github.com/systemd/systemd/pull/13572

However, in the `--console=pipe --as-pid2` use-case, this results in a container with no controlling terminal, which breaks various tools. An in interactive use, it results in `Ctrl-C` in the `systemd-nspawn` container killing the entire container, which is very annoying for shell users.

Before fix:

```
$ sudo systemd-nspawn --console=pipe --as-pid2 --directory=my_chroot
-bash: cannot set terminal process group (-1): Inappropriate ioctl for device
-bash: no job control in this shell
-bash-4.4# cut -d\  -f 7 /proc/$$/stat
0
-bash-4.4# <Ctrl-C here>Container mychroot terminated by signal KILL.
<Afterwards, I need to `reset` in my outer shell to get it to be usable.>
```

After fix:

```
$ sudo ./build/systemd-nspawn --console=pipe --as-pid2 --directory=my_chroot
-bash-4.4# cut -d\  -f 7 /proc/$$/stat
34817
-bash-4.4# ^C
-bash-4.4# ^C
-bash-4.4# exit
logout
```

And the "pipeline" use-case still works fine:

```
$ xxd /dev/zero | sudo ./build/systemd-nspawn --console=pipe --as-pid2 --directory=my_chroot cat | head -n 2
0000000: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000010: 0000 0000 0000 0000 0000 0000 0000 0000  ................
```

poettering reviewed (changes requested):
Hmmm, I am not sure I grok this. So the main use case for `--console=-pipe` is that it is used in shell pipelines (as the name suggests), i.e. where the fd passed through are actually pipe fds, and not ttys, like you appear to use things in. If you invoke nspawn into a shell pipeline we actually want that the whole container go down when the pipeline triggers SIGPIPE or the user hits C-c. it would be strange if used that way and the container wouldn't go down but just the program inside it.

Current behaviour: 

1. if we are asked to do the ptyfwd stuff (i.e. --console=interactive/read-only/passive), pid 1 and pid 2 will have a setsid session of their own each, and pid 2 is going to be the controller of the pty. nspawn itself (i.e. the supervisor process) will remain controller of the calling tty if it is called as that, but it will put the tty in raw mode, so that SIGINT/… can be collected and forwarded to the allocated payload-facing pty.
2. if we are asked to not do the ptyfwd stuff (i.e. --console=pipe) then pid 1 and pid 2 will have a setsid session of their own each too, and neither is a controller of any tty. nspawn itself will remain controller of the calling tty, as before, however it will not be put in raw mode, hence SIGINT/… affect nspawn itself.

I'd argue that this makes a ton of sense.

On top of that: there's a security issue implied by all of that. being a controller of a tty gives you certain powers, hence it's a bit problematic if the payload inside a container (i.e. something substantially less priv'ed) becomes one of a host tty. Becoming one of a payload facing pty is much less problematic, since we isolate it from the host.

Ultimately, you appear to want inetractivity but do not enable the tty mode, which is a bit contradictory. Why? Can you explain the use case? i.e. why use --console=pipe if you want interactivity?

snarkmaster commented on src/nspawn/nspawn-stub-pid1.c:
```diff
@@ -30,7 +30,7 @@ static int reset_environ(const char *new_environment, size_t length) {
 }
 
-int stub_pid1(sd_id128_t uuid) {
+int stub_pid1(sd_id128_t uuid, ConsoleMode arg_console_mode) {
```
Fixed, thanks!

snarkmaster commented on src/nspawn/nspawn-stub-pid1.c:
```diff
@@ -55,18 +55,21 @@ int stub_pid1(sd_id128_t uuid) {
-                setsid();
+                if (arg_console_mode != CONSOLE_PIPE) {
+                        setsid();
+                }
```
Fixed both, thanks!

snarkmaster commented on src/nspawn/nspawn.c:
```diff
@@ -3222,8 +3222,9 @@ static int inner_child(
-        if (setsid() < 0)
+        if (arg_console_mode != CONSOLE_PIPE && setsid() < 0) {
                 return log_error_errno(errno, "setsid() failed: %m");
+        }
```
Fixed. Sorry about the extra `{}`s. This was hardcoded in my brain after Heartbleed, but I'll follow the style guide, of course. Requiring `{}` would best be done as a global, enforced change anyway.

snarkmaster commented:
In a nutshell: we use `systemd-nspawn -aP` as a "POSIXly-transparent chroot+unshare with filesystem setup". We are not going for "container" semantics, but rather for vanilla child process semantics.

The key is the combination of `--as-pid2` with `--console=pipe`. In this setting, the death of the program implies the death of the PID namespace.

I updated the condition in `nspawn.c` to `setsid()` unless **both** flags are set. I also updated the man page and the PR description accordingly. I think this makes more sense now.

Here are more details on why we want to use `--console=pipe` unconditionally: 
 - We only run trusted containers. Your security warning is good. I am familiar with it from the man page under `--console=pipe`. My changes do not affect the security posture -- even if you just use `systemd-nspawn -PD untrusted-filesystem` in a pipeline, you are exposing a TTY on stderr, losing the game.
 - The problem with `isatty(fd)` checks that toggle PTY forwarding is that there is no "correct" value of `fd` -- 0, 1, 2 can all be redirected (or not) in the wild. As a result, if we try to conditionally toggle between `--console=interactive` and `--console=pipe`, we will inflict weird pipeline breakages on our users, and have to spend time debugging them.

In contrast, if we use `-aP` unconditionally, with the present change, both pipelines and interactive usage "just work".

poettering commented:
I am not sure I fully grok your usecase, but wouldn't it be a better fix for your case if we'd hadd --console=auto, which would operate like --console=interactive if stdin/stdout are a TTY, and like --console=pipe otherwise? That way you had a single option you can unconditionall specify and it would result in the right behaviour in all cases. Wouldn't that work?

snarkmaster commented:
@poettering,  the use-case is that we use `systemd-nspawn` as a plain `chroot` with filesystem set-up and process tear-down. Session / TTY management breaks the "vanilla POSIX child process" abstraction.

Your `--console=autopipe` approach is most likely "good enough" but I'm not sure it's 100% right. My concern boils down to the following: PTY forwarding can be enabled unexpectedly, resulting in a real head-scratcher for the developer. I remember being pretty stumped when `systemd-nspawn ... some-script 2> errors` didn't behave as I expected.

That said, if you are unconvinced by this change, I am happy to be pragmatic, and wait until we have some history of edge cases / problems with the `autopipe` approach. 

Also, I recently thought of an (uglier) approach that gets me a "POSIXly transaprent" behavior without changing nspawn:
 - `systemd-nspawn ... sh -c 'exec cat < PIPE_FROM_PARENT > /dev/null'`  -- create the container, wait for pipe close to exit
 - move into the right cgroup, then `nsenter -a -t PID_OF_CAT my_program with args`

This is pretty much what we have to do for the booted `systemd-nspawn` case, so it's not quite as crazy for our setup as it sounds.

Of course, this here PR simply makes `systemd-nspawn` do the right thing for us. I would also challenge you to think of a valid use-case that my PR breaks :)

So, either:
 - let me know if you want me to merge my changes on top of your #17082, or 
 - close this out if you'd like me to come back later with specific user stories to support the "autopipe isn't 100% right" argument.

I don't mind trying it your way.

poettering commented:
#17082 has been merged by now. Let's go with this for the moment, and see how it works out. If it doesn't, please report back, maybe we can tweak things further! Hope that makes sense!

snarkmaster commented:
@poettering, so I tried `autopipe` for our use-case. I'm seeing broken behavior in this setting:
 - Vanilla interactive usage. i.e. the stdin / stdout / stderr of `systemd-nspawn` point at my terminal, such that
 - The nspawn container starts as a subprocess of another program, which manages the container's lifetime, and **logs** to stderr.

While the container runs, the outer program's logs to the same `/dev/pts/N` that it did before starting `systemd-nspawn`. However, the moment that nspawn's PTY forwarding kicks in, (at least) two things break for the parent program (example output below):
 - Newlines stop being interpreted as `\n\r` and become solely `\n` -- see the "staircase logs" below. I can append `\r` to the logs I control, but it's still broken.
 - Worse yet, the host program's FDs 1 & 2 start behave as if they are open in `O_NONBLOCK`, which most code just does not handle -- example error below. Patching all possible code that write to the terminal is not a viable option. Hacks like `dup`ing or reopening the std streams before handing them to `nspawn` do not help because they change something about the terminal flags.

Given that the second issue is prone to causing heisenbugs, and hardening against it is hard, `--console=autopipe` is not really a good choice for us.

So, I want to ask, can we revive this PR? Or are you still worried that making `systemd-nspawn --as-pid2 --console=pipe` act like a regular subprocess, with no terminal shenanigans, will cause problems? Let me know what concerns you have and I'll try to provide evidence one way or another. Thanks!

```
INFO repo_server.py 2020-09-30 21:54:31,507 HTTP `repo-server` is listening
--- Logging error ---
                     Traceback (most recent call last):
                                                         File "/usr/local/fbcode/platform007/lib/python3.7/logging/__init__.py", line 1028, in emit
                                                                       stream.write(msg + self.terminator)
                          BlockingIOError: [Errno 11] write could not complete without blocking
               Call stack:
                            File "<string>", line 38, in <module>
                                                                   File "<string>", line 36, in __run
                       File "/usr/local/fbcode/platform007/lib/python3.7/runpy.py", line 193, in _run_module_as_main
                                        "__main__", mod_spec)
                                                               File "/usr/local/fbcode/platform007/lib/python3.7/runpy.py", line 85, in _run_code
                                                                     exec(code, run_globals)
              File "/data/users/lesha/fbsource/fbcode/buck-out/dev/gen/antlir/rpm/repo-server#link-tree/antlir/rpm/repo_server.py", line 411, in <module>
                                                                             log.info("HTTP `repo-server` is listening")
                                        Message: 'HTTP `repo-server` is listening'
  Arguments: ()
```

PS: Relatedly: I just discovered #17229, which is another way in which `--console=pipe` is currently not POSIXly-transparent. I suspect this one is a more obvious "should fix", and if we can revive the current PR I can even take a stab at fixing the `chown` behavior.

poettering commented:
hmm, my ideas:

1. the fact that we set O_NONBLOCK on stdin/stdout which fucks up other programs is indeed a bit of problem i guess. I figure we should reopen the fds in ptyfwd.c before O_NONBLOCK so that we can set the flag on our own reopened fds, but leave the original fds we are called from untouched. After all we might be killed anytime, and should probably not affect the original fds needlessly then.

2. I'd be fine with adding `--console=as-is` or so (or maybe fine a better name), which does what you want: pass the fds as they are, not create sessions or pgroups or anything. Happy to take a PR for that.
